File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/library/items/media/audio/DiskFileCacheDataSource.java
package com.lasthopesoftware.bluewater.client.library.items.media.audio;

import android.net.Uri;

import com.google.android.exoplayer2.C;
import com.google.android.exoplayer2.upstream.DataSource;
import com.google.android.exoplayer2.upstream.DataSpec;
import com.google.android.exoplayer2.upstream.HttpDataSource;
import com.lasthopesoftware.bluewater.client.library.items.media.files.cached.stream.CacheOutputStream;
import com.lasthopesoftware.bluewater.client.library.items.media.files.cached.stream.supplier.ICacheStreamSupplier;
import com.lasthopesoftware.resources.uri.PathAndQuery;
import com.namehillsoftware.handoff.promises.Promise;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;

import okio.Buffer;


public class DiskFileCacheDataSource implements DataSource {

	private final static Logger logger = LoggerFactory.getLogger(DiskFileCacheDataSource.class);

	private final HttpDataSource defaultHttpDataSource;
	private final ICacheStreamSupplier cacheStreamSupplier;
	private Promise<CacheOutputStream> promisedOutputStream;

	public DiskFileCacheDataSource(HttpDataSource defaultHttpDataSource, ICacheStreamSupplier cacheStreamSupplier) {
		this.defaultHttpDataSource = defaultHttpDataSource;
		this.cacheStreamSupplier = cacheStreamSupplier;
	}

	@Override
	public long open(DataSpec dataSpec) throws IOException {
		if (dataSpec.position == 0) {
			promisedOutputStream =
				cacheStreamSupplier
					.promiseCachedFileOutputStream(PathAndQuery.forUri(dataSpec.uri));
		}

		return defaultHttpDataSource.open(dataSpec);
	}

	@Override
	public int read(byte[] bytes, int offset, int readLength) throws IOException {
		final int result = defaultHttpDataSource.read(bytes, offset, readLength);

		if (promisedOutputStream == null) return result;

		if (result == C.RESULT_END_OF_INPUT) {
			promisedOutputStream
				.eventually(cachedFileOutputStream ->
					cachedFileOutputStream
						.flush()
						.eventually(
							os -> {
								os.close();
								return os.commitToCache();
							},
							e -> {
								logger.warn("An error occurred flushing the output stream", e);
								cachedFileOutputStream.close();
								return Promise.empty();
							}));

			return result;
		}

		final Buffer buffer = new Buffer();
		buffer.write(bytes, offset, result);

		promisedOutputStream = promisedOutputStream
			.eventually(cachedFileOutputStream -> {
				final Promise<CacheOutputStream> promisedWrite = cachedFileOutputStream.promiseTransfer(buffer);
				promisedWrite.then(
					os -> {
						buffer.close();
						return null;
					},
					e -> {
						logger.warn("An error occurred storing the audio file", e);
						buffer.close();
						cachedFileOutputStream.close();
						return null;
					});

				return promisedWrite;
			});

		return result;
	}

	@Override
	public Uri getUri() {
		return defaultHttpDataSource.getUri();
	}

	@Override
	public void close() throws IOException {
		defaultHttpDataSource.close();
	}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/caching/DiskFileCacheDataSource.kt
package com.lasthopesoftware.bluewater.client.playback.caching

import android.net.Uri
import androidx.core.net.toUri
import com.google.android.exoplayer2.C
import com.google.android.exoplayer2.upstream.DataSource
import com.google.android.exoplayer2.upstream.DataSpec
import com.google.android.exoplayer2.upstream.HttpDataSource
import com.google.android.exoplayer2.upstream.TransferListener
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.cached.ICache
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.cached.stream.CacheOutputStream
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.cached.stream.supplier.ICacheStreamSupplier
import com.lasthopesoftware.bluewater.shared.drainQueue
import com.lasthopesoftware.bluewater.shared.promises.NoopResponse.Companion.ignore
import com.lasthopesoftware.bluewater.shared.promises.extensions.toPromise
import com.lasthopesoftware.bluewater.shared.promises.toFuture
import com.lasthopesoftware.resources.uri.PathAndQuery
import com.namehillsoftware.handoff.promises.Promise
import okhttp3.internal.closeQuietly
import okio.Buffer
import org.slf4j.LoggerFactory
import java.io.FileInputStream
import java.io.InputStream
import java.util.concurrent.ConcurrentLinkedQueue

class DiskFileCacheDataSource(
	private val innerDataSource: HttpDataSource,
	private val cacheStreamSupplier: ICacheStreamSupplier,
	private val cachedFilesProvider: ICache
) : DataSource {
	private var inputStream: InputStream? = null
	private var cacheWriter: CacheWriter? = null
	private lateinit var currentDataSpec: DataSpec

	override fun addTransferListener(transferListener: TransferListener) = innerDataSource.addTransferListener(transferListener)

	override fun open(dataSpec: DataSpec): Long {
		cacheWriter?.commit()?.also { cacheWriter = null }
		inputStream?.close()?.also { inputStream = null }

		val key = "${PathAndQuery.forUri(dataSpec.uri)}:${dataSpec.position}:${dataSpec.length}"
		val cachedFile = cachedFilesProvider
			.promiseCachedFile(key)
			.eventually {
				it?.toPromise() ?: cacheStreamSupplier
					.promiseCachedFileOutputStream(key)
					.then { os ->
						cacheWriter = CacheWriter(os)
						it
					}
			}
			.toFuture()
			.get()

		currentDataSpec = dataSpec
		return cachedFile?.let { cf ->
			currentDataSpec = dataSpec.buildUpon()
				.setUri(cf.toUri())
				.setLength(cf.length())
				.build()
			inputStream = FileInputStream(cf)
			cf.length()
		} ?: innerDataSource.open(dataSpec)
	}

	override fun read(bytes: ByteArray, offset: Int, readLength: Int): Int =
		inputStream?.read(bytes, offset, readLength)?.let { if (it != 0) it else C.LENGTH_UNSET }
			?: innerDataSource.read(bytes, offset, readLength).also { result ->
				cacheWriter?.apply {
					if (result != C.RESULT_END_OF_INPUT) queueAndProcess(bytes, offset, result)
					else commit().also { cacheWriter = null }
				}
			}

	override fun getUri(): Uri = currentDataSpec.uri

	override fun getResponseHeaders(): Map<String, List<String>> = innerDataSource.responseHeaders

	override fun close() {
		innerDataSource.close()
		inputStream?.close()
		cacheWriter?.clear()
	}

	private class CacheWriter(private val cachedOutputStream: CacheOutputStream) {
		private val buffersToTransfer = ConcurrentLinkedQueue<Buffer>()
		private val activePromiseSync = Any()

		@Volatile
		private var activePromise = Promise(cachedOutputStream)

		@Volatile
		private var workingBuffer = Buffer()

		@Synchronized
		fun queueAndProcess(bytes: ByteArray, offset: Int, length: Int) {
			workingBuffer.write(bytes, offset, length)
			if (workingBuffer.size < goodBufferSize) return

			buffersToTransfer.offer(workingBuffer)

			workingBuffer = Buffer()

			processQueue().ignore()
		}

		private fun processQueue() : Promise<CacheOutputStream> = synchronized(activePromiseSync) {
			activePromise.eventually({
				if (it == null) return@eventually Promise.empty()

				val concatenatedBuffer = buffersToTransfer
					.drainQueue()
					.reduceOrNull { sink, source -> sink.also(source::readAll) }

				if (concatenatedBuffer == null || concatenatedBuffer.exhausted()) it.toPromise()
				else it.promiseTransfer(concatenatedBuffer)
			}, {
				logger.warn("An error occurred copying the buffer, closing the output stream", it)
				clear()
				Promise.empty()
			}).also { activePromise = it }
		}

		fun commit() {
			synchronized(activePromiseSync) {
				activePromise = activePromise
					.eventually {
						buffersToTransfer.offer(workingBuffer)
						processQueue()
					}
					.eventually { cachedOutputStream.flush() }
					.eventually { cachedOutputStream.commitToCache() }
					.must { cachedOutputStream.close() }
					.then { cachedOutputStream }
			}
		}

		fun clear() {
			synchronized(activePromiseSync) {
				activePromise.must {
					cachedOutputStream.closeQuietly()
					buffersToTransfer.drainQueue().forEach { it.clear() }
				}
			}
		}
	}

	class Factory(
		private val httpDataSourceFactory: HttpDataSource.Factory,
		private val cacheStreamSupplier: ICacheStreamSupplier,
	 	private val cachedFilesProvider: ICache) : DataSource.Factory {
		override fun createDataSource(): DataSource = DiskFileCacheDataSource(
			httpDataSourceFactory.createDataSource(),
			cacheStreamSupplier,
			cachedFilesProvider
		)
	}

	companion object {
		private val logger by lazy { LoggerFactory.getLogger(DiskFileCacheDataSource::class.java) }
		private const val goodBufferSize = 1024L * 1024L // 1MB
	}
}
